File: AlgorithmDataStructure/LeetCode/543-DiameterOfTree/diameter.c

#include <stdio.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


int depth(struct TreeNode *node, int *res) {
    if (node == NULL) { return 0; }
    int left = depth(node->left, res);
    int right = depth(node->right, res);
    int sum = left + right;
    *res = *res < sum ? sum : *res;
    return left < right ? right + 1 : left + 1;
}

int diameterOfBinaryTree(struct TreeNode* root){
    if (root == NULL) { return 0; }
    int res = 0;
    depth(root, &res);
    return res;
}


int main(int argc, const char * argv[]) {
    
    struct TreeNode r1 = {9, NULL, NULL};
    struct TreeNode l1 = {7, NULL, NULL};
    struct TreeNode left = {6, &l1, NULL};
    struct TreeNode right = {5, NULL, &r1};
    
    struct TreeNode t1 = {1, &left, &right};
    
    
    int a = diameterOfBinaryTree(&t1);
    printf("=== %d\n", a);
    return 0;
}

File: AlgorithmDataStructure/LeetCode/543-DiameterOfTree/diameter.swift

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

class Solution {
    var res = 0
    func diameterOfBinaryTree(_ root: TreeNode?) -> Int {
        if root == nil { return 0 }
        deep(root)
        return res
    }
    
    func deep(_ node: TreeNode?) -> Int {
        if node == nil { return 0 }
        let l = deep(node!.left)
        let r = deep(node!.right)
        res = max(res, l + r)
        return max(l, r) + 1
    }
}


var l1 = TreeNode(9)
var left = TreeNode(7)
left.left = l1
var right = TreeNode(6)
var t = TreeNode(1)
t.left = left
t.right = right

let s = Solution()
print(s.diameterOfBinaryTree(t))

